File: app/src/main/java/com/chuongvd/app/signal/widget/adapter/BaseViewHolderBinding.java
package com.chuongvd.app.signal.widget.adapter;

import android.databinding.ViewDataBinding;
import android.support.v7.widget.RecyclerView;

public class BaseViewHolderBinding<BINDINGVIEW extends ViewDataBinding, MODEL>
        extends RecyclerView.ViewHolder {

    public BINDINGVIEW mBinding;
    public BaseRecyclerViewAdapterBinding.OnRecyclerItemListener<MODEL> mListener;
    private MODEL data;

    public BaseViewHolderBinding(BINDINGVIEW binding) {
        super(binding.getRoot());
        mBinding = binding;
    }

    public BaseViewHolderBinding(BINDINGVIEW binding,
            BaseRecyclerViewAdapterBinding.OnRecyclerItemListener<MODEL> listener) {
        super(binding.getRoot());
        mBinding = binding;
        mListener = listener;
    }

    public void bindData(MODEL MODEL) {
        data = MODEL;
        mBinding.getRoot().setOnClickListener(v -> {
            if (mListener != null) {
                mListener.onItemClick(getAdapterPosition(), data);
            }
        });
    }
}

File: app/src/main/java/com/chuongvd/app/signal/widget/adapter/BaseViewHolderBinding.kt
package com.chuongvd.app.signal.widget.adapter

import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView

open class BaseViewHolderBinding<BINDINGVIEW : ViewDataBinding, MODEL> : RecyclerView.ViewHolder {

  var mBinding: BINDINGVIEW
  var mListener: BaseRecyclerViewAdapterBinding.OnRecyclerItemListener<MODEL>? = null
  private var data: MODEL? = null

  constructor(binding: BINDINGVIEW) : super(binding.root) {
    mBinding = binding
  }

  constructor(binding: BINDINGVIEW,
      listener: BaseRecyclerViewAdapterBinding.OnRecyclerItemListener<MODEL>) : super(
      binding.root) {
    mBinding = binding
    mListener = listener
  }

  open fun bindData(model: MODEL) {
    data = model
    mBinding.root.setOnClickListener { v ->
      if (mListener != null) {
        mListener!!.onItemClick(adapterPosition, model)
      }
    }
  }
}
